### ISSUE TITLE
x11 with Xwayland: requesting fullscreen causes BadRRCrtc and makes subsequent SDL programs fail

### ISSUE THREAD

**Issue by smcv** [libsdl-org/SDL#4630]
I'm trying to upgrade the SDL package in Debian from 2.0.14 to the recent 2.0.16 release, and encountered a regression. I'm running GNOME 3.38 in Wayland mode with Xwayland, which will be the default desktop environment in Debian 11 (we can't have GNOME 40 in Debian 11, because that wasn't released until after Debian 11 had already frozen).

When I run a fullscreen SDL program like `testgl2 --fullscreen` or OpenArena using the default X11 video backend for SDL, the updated SDL fails with an X11 error:

```
X Error of failed request:  BadRRCrtc (invalid Crtc parameter)
  Major opcode of failed request:  140 (RANDR)
  Minor opcode of failed request:  20 (RRGetCrtcInfo)
  Crtc id in failed request: 0x0
  Serial number of failed request:  243
  Current serial number in output stream:  243
```

SDL 2.0.16 triggers this, but a modified SDL 2.0.16 with b033cd0d, 4c7825f6, d0effadf and 16e3bfe8 reverted *does not* trigger it.

When this happens, it [changes the state of my Xwayland display](https://github.com/libsdl-org/SDL/issues/4630#issuecomment-896934986), leaving the mode unset.

After the state has changed, running any other graphical SDL program with the X11 backend (for example `testgl2`, or `openarena` again) fails, this time with a different error:

    X Error of failed request:  BadValue (integer parameter out of range for operation)
      Major opcode of failed request:  151 (XFree86-VidModeExtension)
      Minor opcode of failed request:  1 (XF86VidModeGetModeLine)
      Value in failed request:  0x522
      Serial number of failed request:  63
      Current serial number in output stream:  63

After the state has changed, all versions of SDL that I have tried will fail in the same way (I tried vanilla 2.0.16, 2.0.16 + reverts as listed above, and 2.0.14dfsg2-3 from Debian 11), which means it's easy to get the wrong idea about which versions work and which versions don't.

The changed state persists until I log out from GNOME and back in.

<details><summary>Original report (misleading)</summary>

Games like `openarena` and `darkplaces`, and also simpler test programs like `testgl2` and `testvulkan`, fail with:

    X Error of failed request:  BadValue (integer parameter out of range for operation)
      Major opcode of failed request:  151 (XFree86-VidModeExtension)
      Minor opcode of failed request:  1 (XF86VidModeGetModeLine)
      Value in failed request:  0x522
      Serial number of failed request:  63
      Current serial number in output stream:  63

Workaround: if I run with `SDL_VIDEODRIVER=wayland`, the games and test programs work (but without window decorations because we don't have `libdecor` available yet - I'm working on that in parallel).

This seems like might be related to #4561, where @icculus mentioned that XVidMode "doesn't work in SDL2 in almost all cases", but the failure mode is different. I have not done anything special to force Xrandr to be disabled, and I don't have any `SDL_*` in my environment.

I'm building with essentially the same options that we used in the 2.0.14 package, and in particular `configure.ac` is successfully detecting Xrandr.

I tried building 2.0.16 with b033cd0d, 4c7825f6, d0effadf and 16e3bfe8 reverted, but that doesn't seem to resolve this.

</details>

**Comment by icculus** [comment]
Can you post config.log here?

**Comment by smcv** [comment]
Actually, this seems to be a regression somewhere else on the system, rather than a SDL regression: I downgraded to the Debian 11 package (2.0.14+dfsg2-3) and I'm still getting this error. It worked when I uploaded 2.0.14+dfsg2-3 back in January, so there must be something else going wrong.

**Comment by smcv** [comment]
I can't reproduce this after logging out from GNOME and back in, either with 2.0.14 or with 2.0.16.

**Comment by smcv** [comment]
A more verbose version of the state change: before running openarena

```
$ xrandr --verbose
Screen 0: minimum 16 x 16, current 1920 x 1080, maximum 32767 x 32767
XWAYLAND0 connected primary 1920x1080+0+0 (0x22) normal (normal left inverted right x axis y axis) 310mm x 170mm
	Identifier: 0x21
	Timestamp:  233790193
	Subpixel:   unknown
	Gamma:      1.0:1.0:1.0
	Brightness: 0.0
	Clones:    
	CRTC:       0
	CRTCs:      0
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	non-desktop: 0 
		supported: 0, 1
  1920x1080 (0x22) 173.000MHz -HSync +VSync *current +preferred
        h: width  1920 start 2048 end 2248 total 2576 skew    0 clock  67.16KHz
        v: height 1080 start 1083 end 1088 total 1120           clock  59.96Hz
```

and after

```
$ xrandr --verbose
Screen 0: minimum 16 x 16, current 1920 x 1080, maximum 32767 x 32767
XWAYLAND0 connected primary (normal left inverted right x axis y axis)
	Identifier: 0x21
	Timestamp:  234157697
	Subpixel:   unknown
	Clones:    
	CRTCs:      0
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	non-desktop: 0 
		supported: 0, 1
  1920x1080 (0x22) 173.000MHz -HSync +VSync +preferred
        h: width  1920 start 2048 end 2248 total 2576 skew    0 clock  67.16KHz
        v: height 1080 start 1083 end 1088 total 1120           clock  59.96Hz
```

**Comment by smcv** [comment]
I've updated the description to reflect my current understanding of this bug. There are two failure modes, not just one.

The root cause seems to be that when I run openarena with the new SDL, it reconfigures my Xwayland's Xrandr configuration, leaving it in a broken state. I'll try to make a more minimal reproducer for this than all of openarena. This part of the bug only affects 2.0.16, and can be avoided by reverting #4494 and the follow-up commits.

After Xwayland's Xrandr state has been changed like this, all versions of SDL fail to start up - not just 2.0.16, but also older versions. I think this might not be a real bug in those versions: obviously it would be ideal if they could recover, but I haven't been able to work out how to recover from the bad state using `xrandr(1)` either. This can be avoided by making sure not to get Xwayland into the bad state in the first place.

**Comment by smcv** [comment]
I can also reproduce this with `testgl2 --fullscreen`, which is a much simpler reproducer.

**Comment by smcv** [comment]
After logging out and back in to reset the state, `SDL_VIDEO_X11_XRANDR=0 testgl2 --fullscreen` also does not trigger this.

**Comment by bluestang2006** [comment]
this is very similar to the issue I am having with [Quake3e](https://github.com/ec-/Quake3e), however, I am only experiencing this with the Vulkan renderer in Fullscreen mode only (windowed works)...OpenGL works in both Fullscreen and Windowed...

https://github.com/ec-/Quake3e/issues/119

https://gitlab.freedesktop.org/mesa/mesa/-/issues/5207

Perhaps there is something related?

**Comment by smcv** [comment]
This is not GNOME-specific, I get the same thing with KDE Plasma in Wayland mode + Xwayland.

**Comment by smcv** [comment]
However, this *does* seem to be Xwayland-specific: GNOME in Xorg mode on the same hardware and software (AMD GPU, Debian 11) does not have this behaviour, and neither does KDE Plasma in Xorg mode.

**Comment by smcv** [comment]
Unlike #4561, what I'm seeing here is that the *second* call to `X11_XRRSetCrtcConfig()` (the one that switches the crtc back on) fails, with result 3 (`BadWindow`) if I'm reading correctly. So the call sequence goes something like this:

* get the previous crtc info
* grab server
* turn off crtc (succeeds)
* set screen size, with workarounds for #4561 
* turn crtc back on (fails!)

and then the crtc is left in a disabled state (because we couldn't turn it back on) until I exit and restart Xwayland.

**Comment by smcv** [comment]
Applying either #4631 or #4632 seems to avoid this issue.

(There might be a way to solve it that's better than either of those, but I don't know X11 well enough to see it.)

**Comment by smcv** [comment]
I can reproduce something similar to this without SDL:

```
$ xrandr --output XWAYLAND0 --mode 1920x1080
(ok)
$ xrandr --output XWAYLAND0 --pos 0x0 --mode 1920x1080
(ok)
$ xrandr --output XWAYLAND0 --crtc 0 --pos 0x0 --mode 1920x1080
(ok)
$ xrandr --output XWAYLAND0 --crtc 0 --pos 0x0 --mode 1920x1080 --rotate inverted
xrandr: Configure crtc 0 failed
```

and, similar to the situation with SDL, the XWAYLAND0 virtual display is no longer connected to a virtual CRTC - no mode is listed as `*current` in `xrandr --verbose`, and `CRTC: 0` no longer appears.

After I have got Xwayland into this state, simple X11 apps like xeyes still work, but SDL's `testgl2` does not (even when run without `--fullscreen`).

**Comment by bluestang2006** [comment]
> Applying either #4631 or #4632 seems to avoid this issue.
> 
> (There might be a way to solve it that's better than either of those, but I don't know X11 well enough to see it.)

Neither of those PRs work for me. Same behavior as I described before where the Vulkan renderer is the one with the issue in Fullscreen mode only. 

I am using Ubuntu 21.04, so it's definitely an XWayland issue...I tried logging in with just XOrg and didn't have issues with the Vulkan renderer.

**Comment by smcv** [comment]
@bluestang2006, if #4631 (which is just a revert) doesn't work for you, then I think you are experiencing a different bug and should open a separate issue.

**Comment by bluestang2006** [comment]
> @bluestang2006, if #4631 (which is just a revert) doesn't work for you, then I think you are experiencing a different bug and should open a separate issue.

You are right, it's a different bug...forcing `MESA_VK_WSI_PRESENT_MODE=fifo` or `MESA_VK_WSI_PRESENT_MODE=relaxed` fixes my issue w/ the Vulkan renderer and Fullscreen mode.

However, the console output still spits out the present mode as immediate...

**Comment by smcv** [comment]
I've opened https://gitlab.freedesktop.org/xorg/xserver/-/issues/1209 to query whether this is an Xwayland bug. However, I'd like to get a workaround into SDL even if this is fixed in Xwayland, because Xwayland is part of the OS (hence long-term frozen in some distributions), but SDL is frequently bundled with Steam or games.

**Comment by smcv** [comment]
The underlying Xwayland bug is fixed in the [standalone Xwayland 21.x releases](https://fedoraproject.org/wiki/Changes/XwaylandStandalone) used in e.g. Fedora, Arch and very recent Ubuntu (but not Debian yet).
